File: scripts/set_supabase_claim.py
"""
Server-side Python script to assign { role: "authenticated" } claim to Firebase users for Supabase RLS.

Usage:
  python scripts/set_supabase_claim.py <FIREBASE_UID_OR_PHONE> [path_to_service_account.json]
  python scripts/set_supabase_claim.py --all [path_to_service_account.json]

Requirements:
  pip install firebase-admin
"""

import sys
import os
import firebase_admin
from firebase_admin import auth, credentials

def initialize_app(service_account_path=None):
    if firebase_admin._apps:
        return
    if service_account_path and os.path.exists(service_account_path):
        cred = credentials.Certificate(service_account_path)
        firebase_admin.initialize_app(cred)
    elif os.environ.get("GOOGLE_APPLICATION_CREDENTIALS") and os.path.exists(os.environ["GOOGLE_APPLICATION_CREDENTIALS"]):
        firebase_admin.initialize_app()
    else:
        try:
            firebase_admin.initialize_app(options={"projectId": "sih2026-75333"})
        except Exception as e:
            print(f"Error initializing Firebase Admin SDK: {e}")
            print("Please provide the service account JSON path as argument or set GOOGLE_APPLICATION_CREDENTIALS.")
            sys.exit(1)

def assign_claim(user_record):
    current_claims = user_record.custom_claims or {}
    updated_claims = {**current_claims, "role": "authenticated"}
    auth.set_custom_user_claims(user_record.uid, updated_claims)
    print(f"[SUCCESS] Assigned {{'role': 'authenticated'}} to UID: {user_record.uid} ({user_record.phone_number or 'no-phone'})")

def main():
    if len(sys.argv) < 2:
        print("Usage:")
        print("  python scripts/set_supabase_claim.py <FIREBASE_UID_OR_PHONE> [path_to_service_account.json]")
        print("  python scripts/set_supabase_claim.py --all [path_to_service_account.json]")
        sys.exit(1)

    target = sys.argv[1]
    sa_path = sys.argv[2] if len(sys.argv) > 2 else None
    initialize_app(sa_path)

    if target == "--all":
        page = auth.list_users()
        count = 0
        while page:
            for u in page.users:
                assign_claim(u)
                count += 1
            page = page.get_next_page()
        print(f"Finished assigning claims to {count} users.")
        return

    if target.startswith("+"):
        user = auth.get_user_by_phone_number(target)
    else:
        user = auth.get_user(target)

    assign_claim(user)
    refreshed = auth.get_user(user.uid)
    print("Verified claims:", refreshed.custom_claims)

if __name__ == "__main__":
    main()

File: scripts/set_supabase_claim.js
/**
 * Server-side script to assign Supabase-compatible custom claim to Firebase users.
 *
 * Sets { role: "authenticated" } so that Supabase Third-Party Auth recognizes
 * Firebase JWTs as the PostgreSQL "authenticated" role for Storage RLS.
 *
 * Usage:
 *   node scripts/set_supabase_claim.js <FIREBASE_UID_OR_PHONE>
 *   node scripts/set_supabase_claim.js --all
 *
 * Requirements:
 *   npm install firebase-admin
 *   Set GOOGLE_APPLICATION_CREDENTIALS=/path/to/serviceAccountKey.json
 *   (or pass serviceAccountKey.json path as second argument)
 */

const { initializeApp, getApps, cert, applicationDefault } = require('firebase-admin/app');
const { getAuth } = require('firebase-admin/auth');
const path = require('path');
const fs = require('fs');

// Initialize Firebase Admin SDK
if (!getApps().length) {
  const serviceAccountArg = process.argv[3];
  if (serviceAccountArg && fs.existsSync(serviceAccountArg)) {
    const serviceAccount = JSON.parse(fs.readFileSync(path.resolve(serviceAccountArg), 'utf8'));
    initializeApp({
      credential: cert(serviceAccount),
    });
  } else if (process.env.GOOGLE_APPLICATION_CREDENTIALS && fs.existsSync(process.env.GOOGLE_APPLICATION_CREDENTIALS)) {
    initializeApp({
      credential: applicationDefault(),
    });
  } else {
    try {
      initializeApp({
        projectId: 'sih2026-75333',
      });
    } catch (e) {
      console.error('Failed to initialize Firebase Admin SDK. Please set GOOGLE_APPLICATION_CREDENTIALS or provide the service account JSON path as an argument.');
      process.exit(1);
    }
  }
}

const auth = getAuth();

async function setClaimForUser(userRecord) {
  const existingClaims = userRecord.customClaims || {};
  const updatedClaims = {
    ...existingClaims,
    role: 'authenticated',
  };

  await auth.setCustomUserClaims(userRecord.uid, updatedClaims);
  console.log(`[SUCCESS] Assigned { role: 'authenticated' } to UID: ${userRecord.uid} (${userRecord.phoneNumber || userRecord.email || 'no-contact'})`);
}

async function main() {
  const target = process.argv[2];

  if (!target) {
    console.error('Usage:');
    console.error('  node scripts/set_supabase_claim.js <FIREBASE_UID_OR_PHONE>');
    console.error('  node scripts/set_supabase_claim.js --all');
    console.error('  node scripts/set_supabase_claim.js <UID> /path/to/serviceAccountKey.json');
    process.exit(1);
  }

  try {
    if (target === '--all') {
      console.log('Listing all users to assign role=authenticated...');
      let pageToken;
      let total = 0;
      do {
        const listUsersResult = await auth.listUsers(1000, pageToken);
        for (const userRecord of listUsersResult.users) {
          await setClaimForUser(userRecord);
          total++;
        }
        pageToken = listUsersResult.pageToken;
      } while (pageToken);
      console.log(`Done! Assigned claims to ${total} users.`);
      return;
    }

    let user;
    if (target.startsWith('+')) {
      // Lookup by phone number
      user = await auth.getUserByPhoneNumber(target);
    } else {
      // Lookup by UID
      user = await auth.getUser(target);
    }

    await setClaimForUser(user);
    const refreshed = await auth.getUser(user.uid);
    console.log('[VERIFIED] Custom claims:', refreshed.customClaims);
  } catch (error) {
    console.error('[ERROR]', error);
    process.exit(1);
  }
}

main();

